lib/onelogin/ruby-saml/utils.rb

```diff
@@ -15,7 +15,21 @@ class Utils
 
       DSIG      = "http://www.w3.org/2000/09/xmldsig#"
       XENC      = "http://www.w3.org/2001/04/xmlenc#"
-      DURATION_FORMAT = %r(^(-?)P(?:(?:(?:(\d+)Y)?(?:(\d+)M)?(?:(\d+)D)?(?:T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?)?)|(?:(\d+)W))$)
+      DURATION_FORMAT = %r(^
+        (-?)P                         # 1: Duration sign
+        (?:
+          (?:(\d+)Y)?                 # 2: Years
+          (?:(\d+)M)?                 # 3: Months
+          (?:(\d+)D)?                 # 4: Days
+          (?:T
+            (?:(\d+)H)?               # 5: Hours
+            (?:(\d+)M)?               # 6: Minutes
+            (?:(\d+(?:\.\d{3})?)S)?   # 7: Seconds and optional milliseconds
+          )?
+          |
+          (\d+)W                      # 8: Weeks
+        )
+      $)x
```

Comment:
what is this x?

Reply:
That's the "ignore whitespace" regexp flag. It's what allows me to split the regexp over multiple lines, and also insert comments. You can see it here documented for 2.6, but it's supported back to 1.8.x:

https://ruby-doc.org/core-2.6.4/Regexp.html#class-Regexp-label-Options